### PR TITLE
CRM-21227 Fix test following merge of PR #10435

### PR DESCRIPTION
**Pull request by seamuslee001**
Overview
----------------------------------------
Test has started to fail i believe due to PR #10435 see https://test.civicrm.org/job/CiviCRM-Core-PR/17314/testReport/(root)/CRM_Core_Page_HookTest/testFormsCallBuildFormOnce/ as an example

Before
----------------------------------------
Test fails

After
----------------------------------------
Test passes

Technical Details
----------------------------------------
The PR made some changes to how `$defaults` is populated from requests which has made some cascading effects

ping @eileenmcnaughton @monishdeb

---

 * [CRM-21227: Fix issues in CRM_Core_Page_run test suite following merge of PR \#10435](https://issues.civicrm.org/jira/browse/CRM-21227)

### CODE REVIEW

File: CRM/Admin/Form/RelationshipType.php [hunk 1]
```diff
@@ -117,7 +117,7 @@ public function setDefaultValues() {
         $defaults['contact_types_a'] .= '__' . $defaults['contact_sub_type_a'];
       }
 
-      $defaults['contact_types_b'] = $defaults['contact_type_b'];
```
Comment (seamuslee001) [inline CRM/Admin/Form/RelationshipType.php]: This is just changing to be the same as L115 but for contact_type_b

File: CRM/Badge/Form/Layout.php [hunk 2]
```diff
@@ -154,7 +154,7 @@ public function buildQuickForm() {
   public function setDefaultValues() {
     if (isset($this->_id)) {
       $defaults = array_merge($this->_values,
-        CRM_Badge_BAO_Layout::getDecodedData($this->_values['data']));
+        CRM_Badge_BAO_Layout::getDecodedData(CRM_Utils_Array::value('data', $this->_values, '[]')));
```
Comment (seamuslee001) [inline CRM/Badge/Form/Layout.php]: This checks for data in $this->_values and passes an empty json_encoded array as the default value (getDecodedData does a jsonDecode

File: CRM/Contact/Form/DedupeRules.php [hunk 3]
```diff
@@ -60,8 +60,8 @@ public function preProcess() {
     $this->_rgid = CRM_Utils_Request::retrieve('id', 'Positive', $this, FALSE, 0);
     $contactTypes = civicrm_api3('Contact', 'getOptions', array('field' => "contact_type"));
     $contactType = CRM_Utils_Request::retrieve('contact_type', 'String', $this, FALSE, 0);
-    if (in_array($contactType, $contactTypes['values'])) {
-      $this->_contactType = $contactTypes['values'][$contactType];
+    if (CRM_Utils_Array::value($contactType, $contactTypes['values'])) {
```
Comment (mickadoo) [inline CRM/Contact/Form/DedupeRules.php]: Why the change here? It changes the logic from asking "is $contactType in the array $values" to "is $values[$contactType] truthy?"

File: CRM/Financial/Form/FinancialBatch.php [hunk 4]
```diff
@@ -83,8 +83,7 @@ public function preProcess() {
   public function buildQuickForm() {
     parent::buildQuickForm();
     $this->setPageTitle(ts('Financial Batch'));
-
-    if (isset($this->_id)) {
+    if (!empty($this->_id)) {
```
Comment (eileenmcnaughton) [inline CRM/Financial/Form/FinancialBatch.php]: this is fine - although I think it would be better to define the property on the class

File: CRM/Contact/Form/DedupeRules.php (hunk 3 above)
Comment (seamuslee001) [reply CRM/Contact/Form/DedupeRules.php]: @mickadoo Michael if you look at how old L64 was crafted it was using it as a Key whereas in_array checks on the value so somewhere we stuffed up. I'm not sure if should be on the key or not

Comment (eileenmcnaughton) [reply CRM/Contact/Form/DedupeRules.php]: so the effect here is that IF contactType is passed in via the URL  we accept it if valid. If not we calculate from the rule perhaps keep this & change lower down

Remove
    elseif (!empty($contactType)) {
      throw new CRM_Core_Exception('Contact Type is Not valid');
    }

& replace with 

    if (!$this->_rgid && !$this->_contactType) {
      CRM_Core_Session::setStatus('Contact Type could not be determined, assuming Individual');
     $this->_contactType = 'Individual';
}

Or we could throw an exception - but I can't see why we don't just use a default

**Comment by JoeMurray** [comment]
Jenkins, please test this

**Comment by JoeMurray** [comment]
Jenkins, test this please

**Comment by eileenmcnaughton** [comment]
I'm going to merge this because it does fix some tests & I don't think the question about contactTypes is a blocker
